File: src/main/java/com/projeto/springapi/controller/EnderecoController.java
package com.projeto.springapi.controller;

import com.projeto.springapi.dto.EnderecoDTO;
import com.projeto.springapi.model.Endereco;
import com.projeto.springapi.request.EnderecoRequest;
import com.projeto.springapi.service.EnderecoService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/enderecos")
@Tag(name = "Endereços", description = "Operações relacionadas a endereços")
public class EnderecoController {

    @Autowired
    private EnderecoService enderecoService;

    @PostMapping
    @Operation(
            summary = "Cria um novo endereço (utilizando EnderecoRequest)",
            description = "Cria um novo registro de endereço com base nos dados fornecidos no corpo da requisição.",
            requestBody = @io.swagger.v3.oas.annotations.parameters.RequestBody(
                    description = "Dados do endereço a serem criados",
                    required = true,
                    content = @Content(schema = @Schema(implementation = EnderecoRequest.class))
            ),
            responses = {
                    @ApiResponse(responseCode = "201", description = "Endereço criado com sucesso",
                            content = @Content(mediaType = "application/json",
                                    schema = @Schema(implementation = Endereco.class))),
                    @ApiResponse(responseCode = "400", description = "Dados de requisição inválidos")
            }
    )
    public ResponseEntity<Endereco> createEndereco(@RequestBody EnderecoRequest enderecoRequest) {
        Endereco createdEndereco = enderecoService.createEndereco(enderecoRequest);
        return new ResponseEntity<>(createdEndereco, HttpStatus.CREATED);
    }

    @GetMapping("/{id}")
    @Operation(
            summary = "Busca um endereço por ID",
            description = "Retorna os detalhes de um endereço específico com base no ID fornecido.",
            parameters = {
                    @Parameter(name = "id", description = "ID do endereço a ser buscado", required = true, schema = @Schema(type = "integer"))
            },
            responses = {
                    @ApiResponse(responseCode = "200", description = "Endereço encontrado",
                            content = @Content(mediaType = "application/json",
                                    schema = @Schema(implementation = Endereco.class))),
                    @ApiResponse(responseCode = "404", description = "Endereço não encontrado")
            }
    )
    public ResponseEntity<Endereco> getEnderecoById(@PathVariable Integer id) {
        Endereco endereco = enderecoService.getEnderecoById(id);
        return ResponseEntity.ok(endereco);
    }

    @PostMapping
    @Operation(
            summary = "Cria um novo endereço (utilizando EnderecoDTO)",
            description = "Cria um novo registro de endereço com base nos dados fornecidos no corpo da requisição (DTO).",
            requestBody = @io.swagger.v3.oas.annotations.parameters.RequestBody(
                    description = "Dados do endereço a serem criados (DTO)",
                    required = true,
                    content = @Content(schema = @Schema(implementation = EnderecoDTO.class))
            ),
            responses = {
                    @ApiResponse(responseCode = "201", description = "Endereço criado com sucesso",
                            content = @Content(mediaType = "application/json",
                                    schema = @Schema(implementation = EnderecoDTO.class))),
                    @ApiResponse(responseCode = "400", description = "Dados de requisição inválidos")
            }
    )
    public ResponseEntity<EnderecoDTO> createEndereco(@RequestBody EnderecoDTO enderecoDTO) {
        return ResponseEntity.status(HttpStatus.CREATED).body(enderecoService.createEndereco(enderecoDTO));
    }

    @PutMapping("/{id}")
    @Operation(
            summary = "Atualiza um endereço existente",
            description = "Atualiza os dados de um endereço existente com base no ID fornecido e nos dados do corpo da requisição.",
            parameters = {
                    @Parameter(name = "id", description = "ID do endereço a ser atualizado", required = true, schema = @Schema(type = "integer"))
            },
            requestBody = @io.swagger.v3.oas.annotations.parameters.RequestBody(
                    description = "Novos dados para o endereço",
                    required = true,
                    content = @Content(schema = @Schema(implementation = EnderecoDTO.class))
            ),
            responses = {
                    @ApiResponse(responseCode = "200", description = "Endereço atualizado com sucesso",
                            content = @Content(mediaType = "application/json",
                                    schema = @Schema(implementation = EnderecoDTO.class))),
                    @ApiResponse(responseCode = "400", description = "Dados de requisição inválidos"),
                    @ApiResponse(responseCode = "404", description = "Endereço não encontrado")
            }
    )
    public ResponseEntity<EnderecoDTO> updateEndereco(@PathVariable Long id, @RequestBody EnderecoDTO enderecoDTO) {
        return ResponseEntity.ok(enderecoService.updateEndereco(id, enderecoDTO));
    }

    @DeleteMapping("/{id}")
    @Operation(
            summary = "Exclui um endereço por ID",
            description = "Exclui um registro de endereço com base no ID fornecido.",
            parameters = {
                    @Parameter(name = "id", description = "ID do endereço a ser excluído", required = true, schema = @Schema(type = "integer"))
            },
            responses = {
                    @ApiResponse(responseCode = "204", description = "Endereço excluído com sucesso"),
                    @ApiResponse(responseCode = "404", description = "Endereço não encontrado")
            }
    )
    public ResponseEntity<Void> deleteEndereco(@PathVariable Long id) {
        enderecoService.deleteEndereco(id);
        return ResponseEntity.noContent().build();
    }
}

File: app/Http/Controllers/EnderecoController.php
<?php

namespace App\Http\Controllers;

use App\Models\Endereco;
use App\Http\Resources\EnderecoResource;
use Illuminate\Http\Request;

/**
 * @OA\Schema(
 *     schema="Endereco",
 *     type="object",
 *     @OA\Property(property="end_id", type="integer", readOnly=true),
 *     @OA\Property(property="end_tipo_logradouro", type="string"),
 *     @OA\Property(property="end_logradouro", type="string"),
 *     @OA\Property(property="end_numero", type="integer"),
 *     @OA\Property(property="end_bairro", type="string"),
 *     @OA\Property(property="cid_id", type="integer")
 * )
 * @OA\Tag(
 *     name="Endereços",
 *     description="Operações relacionadas a endereços"
 * )
 */
class EnderecoController extends Controller
{
    /**
     * @OA\Get(
     *      path="/api/enderecos/{id}",
     *      operationId="getEndereco",
     *      tags={"Endereços"},
     *      summary="Busca endereço por ID",
     *      security={{"sanctum": {}}},
     *      @OA\Parameter(
     *          name="id",
     *          in="path",
     *          required=true,
     *          @OA\Schema(type="integer")
     *      ),
     *      @OA\Response(response=200, description="Endereço encontrado"),
     *      @OA\Response(response=404, description="Endereço não encontrado")
     * )
     */
    public function show($id)
    {
        $endereco = Endereco::findOrFail($id);
        return new EnderecoResource($endereco);
    }

    /**
     * @OA\Post(
     *      path="/api/enderecos",
     *      operationId="createEndereco",
     *      tags={"Endereços"},
     *      summary="Cria novo endereço",
     *      security={{"sanctum": {}}},
     *      @OA\RequestBody(
     *          required=true,
     *          @OA\JsonContent(ref="#/components/schemas/Endereco")
     *      ),
     *      @OA\Response(response=201, description="Endereço criado"),
     *      @OA\Response(response=400, description="Dados inválidos")
     * )
     */
    public function store(Request $request)
    {
        $data = $request->validate([
            'end_tipo_logradouro' => 'required|string',
            'end_logradouro' => 'required|string',
            'end_numero' => 'required|integer',
            'end_bairro' => 'required|string',
            'cid_id' => 'required|exists:cidade,cid_id'
        ]);

        $endereco = Endereco::create($data);
        return new EnderecoResource($endereco);
    }

    /**
     * @OA\Put(
     *      path="/api/enderecos/{id}",
     *      operationId="updateEndereco",
     *      tags={"Endereços"},
     *      summary="Atualiza endereço",
     *      security={{"sanctum": {}}},
     *      @OA\Parameter(
     *          name="id",
     *          in="path",
     *          required=true,
     *          @OA\Schema(type="integer")
     *      ),
     *      @OA\RequestBody(
     *          required=true,
     *          @OA\JsonContent(ref="#/components/schemas/Endereco")
     *      ),
     *      @OA\Response(response=200, description="Endereço atualizado"),
     *      @OA\Response(response=400, description="Dados inválidos"),
     *      @OA\Response(response=404, description="Endereço não encontrado")
     * )
     */
    public function update(Request $request, $id)
    {
        $endereco = Endereco::findOrFail($id);
        
        $data = $request->validate([
            'end_tipo_logradouro' => 'required|string',
            'end_logradouro' => 'required|string',
            'end_numero' => 'required|integer',
            'end_bairro' => 'required|string',
            'cid_id' => 'required|exists:cidade,cid_id'
        ]);

        $endereco->update($data);
        return new EnderecoResource($endereco);
    }

    /**
     * @OA\Delete(
     *      path="/api/enderecos/{id}",
     *      operationId="deleteEndereco",
     *      tags={"Endereços"},
     *      summary="Exclui endereço",
     *      security={{"sanctum": {}}},
     *      @OA\Parameter(
     *          name="id",
     *          in="path",
     *          required=true,
     *          @OA\Schema(type="integer")
     *      ),
     *      @OA\Response(response=204, description="Endereço excluído"),
     *      @OA\Response(response=404, description="Endereço não encontrado")
     * )
     */
    public function destroy($id)
    {
        $endereco = Endereco::findOrFail($id);
        $endereco->delete();
        return response()->noContent();
    }
}
